lib/esbonio/esbonio/lsp/sphinx/__init__.py

```diff
@@ -252,6 +296,24 @@ def build(self):
             },
         )
 
+    def cb_env_before_read_docs(self, app, env, docnames: List[str]):
+        """Callback handling env-before-read-docs event."""
+        CustomHTMLTranslator.progress = 0
+        CustomHTMLTranslator.docnames = docnames
+        CustomHTMLTranslator.doccount = len(docnames)
+        # comm.log(f"cb_env_before_read_docs {docnames}")
+        # add our edited file to inject content in source-read, even if not physically changed
+        if self.is_preview & (self.preview_docname not in docnames):
+            # don't care about any other docs
+            docnames.clear()
+            docnames.append(self.preview_docname)
+
+    def cb_source_read(self, app, docname, source):
+        """Callback handling source_read event."""
+        if self.is_preview & (docname == self.preview_docname):
+            # log(f"Injecting live-preview content for {docname}")
+            source[0] = self.preview_content
```

Comment:
> Rather than doing this for only the preview document, is there any reason why we shouldn't override the source with the contents of the workspace for all files? 

Because we trigger onChangeEvent there won't really be any unbuilt changes in other files. They will already have been built when that change occured. 

Hypothetically: If we had multiple unsaved files which were not already built, then the handling of all the unsaved changes and keeping track of all the unsaved changes since last built might also be tricky.

Additionally, it is probably out of scope of a previewer to deal with files besides the current active file.

> As well enabling diagnostics etc. for multiple unsaved files

Like linting and stuff? We are currently not doing too much in that regard. We currently don't know enough about the requirements for that.

> it would remove the need to treat triggering builds for unsaved files as special and we could use the standard build() method

save and trigger_sphinx_build_for_usaved_file are pretty much both just call .build(). Our trigger primes the callbacks we have with the necessary information.

We somehow need to know which is the active document and whether we need to inject unsaved content into the build process. For now, we pass the file path in to trigger_sphinx_build_for_usaved_file.

We could possibly create a new generic build trigger that takes multiple arguments, like:
- should build unsaved files
- list of unsaved files. 

But does that actually change anything? We will still have a onSave function (for lsp) and a custom function that can be called when unsaved files need to be built.

Reply:
I made some changes in boschglobal/esbonio#1 which hopefully provides some more context behind the original suggestion